### ISSUE TITLE
Kafka Reactive Receiver poll new messages even already polled are still in processing

### ISSUE THREAD

**Issue by abialas** [reactor/reactor-kafka#228]
I have a Kafka reactive receiver which read from topic and process messages in parallel (with flatMap).

```java
    reactiveConsumer
        .receive()
        .flatMap(receivedRecord -> Mono.delay(Duration.ofMillis(RandomUtils.nextLong(0, 1000)))
            .doOnNext(delay -> log.info("Processing event {}", receivedRecord.value()))
            .then())
        .repeat()
        .subscribe();
``` 
The new events are coming constantly means even if I have not finished processing the previous polled messages the new one already comes in. 
Additional question I have here: if I am processing messages in parallel how should I do the ack / commit? I cannot do it on each message because there is a huge possibility that I will commit higher offset before the lower one (as messages are processed in parallel).

## Expected Behavior
I would expect that new messages will be polled only in case the ones I have polled before have been processed without error.

## Actual Behavior
The new messages are coming constantly means even if I have not finished processing the previous polled messages the new ones already comes in.

**Comment by garyrussell** [comment]
What version are you using? Backpressure support was added in 1.3.3; the current version is 1.3.4.

You can't use manual commits if you go async, for the reason you state, unless you keep track of the offsets.

**Comment by abialas** [comment]
I am using 1.3.3

**Comment by abialas** [comment]
If I go async and will not do commits from which offset I will read after e.g. crash?

**Comment by garyrussell** [comment]
Whatever the last auto committed offset was - async consuming from a partition is fraught with difficulty. It is better to increase the number of partitions to increase concurrency.

**Comment by abialas** [comment]
Is it possible to commit offset when I read the whole "package" I polled? And why I am getting new messages if old ones are not yet processed? Is it intended behaviour?

**Comment by garyrussell** [comment]
Here's how to apply back-pressure

```java
receiver.receive()
		.publishOn(Schedulers.boundedElastic(), 10)
		.concatMap(rr -> Mono.delay(Duration.ofMillis(random.nextInt(1000)))
			.doOnNext(delay -> log.info(rr.value() + "@" + rr.offset()))
			.then())
		.subscribe();
```

This (prefetch in `concatMap`) only allows 10 in-flight messages at a time. You can see the consumer being paused and resumed if you enable debug logging for reactor.kafka. I am not sure why, but reactor reduces this to 8 (`10 - 10 >> 2`) after the first batch. Another poll is done in the background so the next batch can be processed immediately when the current prefetched records are processed.

There is currently no built-in support to know when all records for batch have been consumed. You have to keep track and acknowledge or commit the highest offset for each partition when you know there are no gaps for unprocessed records.

I opened a new issue:

https://github.com/reactor/reactor-kafka/issues/229

**Comment by abialas** [comment]
I would expect (not sure if correct) that when the message is not successfully processed yet then the next poll should not happen (so basically I should move forward with a message reading only in case I processed correctly or send to dlq or similar or just ack / commit offset). Here we are still polling even though messages are being processed. Does it mean the Kafka receiver does not wait for offset commit and move forward? I know that it would be slow (I was used to this behaviour based on the old StreamListener in Spring Cloud Streams with Kafka)

Regarding the performance - flatMap is really quick, much quicker here than concatMap. Should I even consider using flatMap and process messages concurrently or is it an anti-pattern?

Increasing the number of partitions will not help as I have 8 partitions already but I tested with messages sent to the same partition (the same key).

**Comment by garyrussell** [comment]
There is no relationship between the committed offset and consumer position (except on the first poll). The consumer keeps fetching in the background. When we enable back pressure, we pause the consumer so no more records are fetched, but will still poll the consumer so that a rebalance doesn't occur. When the records are processed we resume the consumer so he fetch more records.

I don't really know; I used `concatMap` because of this https://stackoverflow.com/questions/57296097/how-does-backpressure-work-in-project-reactor. It looks like `flatMap()` works too.

```java
.flatMap(rr -> Mono.delay(Duration.ofMillis(random.nextInt(1000)))
	.doOnNext(delay -> log.info(rr.value() + "@" + rr.offset()))
	.then(), 10, 10)
```

and it does seem faster.

We recently added support for out-of-order commits in Spring for Apache Kafka; I opened that issue to see if we can do something similar here. Unless or until that happens, you are on your own for tracking when to commit the offset(s).

**Comment by garyrussell** [comment]
In general, however, increasing partitions/concurrency (multiple receivers) is better/easier than trying to process records from a single partition asynchronously.

**Comment by abialas** [comment]
When I have a huge load which comes to the single partition for the same key I am afraid increasing the number of partition will not help. Can I have multiple receivers for the same partition assigned (same consumer instance)?

What I am thinking about is to you backpressure with flatMap so messages are processed quickly and even if I do the ack for the latest one from the package first I will not lose so much data.

**Comment by garyrussell** [comment]
No; you can't have multiple receivers on the same partition.

You don't  have access to the "package" (`ConsumerRecords`), the `receiver.receive()` applies a `flatMapIterable` to it; you only see the individual records.

Another solution might be a new method on the receiver `receiveAll()` where you'd get the whole batch in one go; then acknowledge the whole batch with one call (but you'd still have to make sure you ack them in the right order - but that would be much easier than dealing with individual records).

I will look into it next week.

**Comment by abialas** [comment]
That would be great. I am just looking at implementation and just for curiosity - you create Flux (from iterable) in receive, how is that possible that this flux is not completed after I read the first "package"?

**Comment by garyrussell** [comment]
The flux is never completed. The first Flux is `Flux<ConsumerRecords>`. `ConsumerRecords` is `Iterable<ConsumerRecord>`. The `flatMapIterable` effectively flattens the contents of every `ConsumerRecords` (returned by the polls) into a single `Flux<ConsumerRecord>`.

When the `FlattenIterableSubscriber` has a current batch that is being iterated over, it doesn't request any more from the upstream Flux, so the evenyt loop pauses the consumer. When the iterator is drained, the `FlattenIterableSubscriber` requests another `ConsumerRecords` so, on the next poll, the consumer is resumed.
